### DISCUSSION TITLE
Small-file performance from distant site

### DISCUSSION THREAD

**Discussion by inkdot7** [moosefs/moosefs#487] [Q&A]
I am running a small moosefs for test at two remote locations, call them A and B.  The roundtrip time (ping) between them is 25 ms.
mfsmaster runs at location A, while I have a chunkservers at both places, with labels `A` and `B`, respectively.

I have created three storage classes, that stores chunks either on `A`, `B` or `A,B` (i.e. last one duplicated at both sites).
I mount with `mfsmount /mfs mfspreflabels=X`, where `X` is `A` or `B` to prefer local data access on the two nodes.

Performance with no tuning at the two sites for `find X -type f | time xargs md5sum | wc` where X is a directory tree with 770 files using one of the three storage classes:

| dir  | read from node A | read from node B |
| ---- | ----------------------- | ----------------------- |
| A   | 0.4  ms/file           | 67 ms/file              |
| B   | 28  ms/file            | 38 ms/file             |
| AB | 0.4 ms/file           | 39 ms/file            |

Changing in `mfsmaster.cfg` to `ATIME_MODE = 4` has no effect:

| dir  | read from node A | read from node B |
| ---- | ----------------------- | ----------------------- |
| A   |  0.4 ms/file           | 67 ms/file              |
| B   |  28 ms/file            | 39 ms/file             |
| AB |  0.4 ms/file           |  39 ms/file            |

Adding the options `-o mfsattrcacheto=100 -o mfsentrycacheto=100 -o mfsdirentrycacheto=100 -o mfsnegentrycacheto=100 -o mfsgroupscacheto=100` to the mount commands also had little effect:

| dir  | read from node A | read from node B |
| ---- | ----------------------- | ----------------------- |
| A   |  0.4 ms/file           | 74 ms/file              |
| B   |  28  ms/file            | 47  ms/file             |
| AB |  0.4 ms/file           |  48 ms/file            |

From the above numbers, it looks like accessing a file from a host costs about **1.5 network round-trip** of the **time distance to `mfsmaster`**.  There is also a cost of about **1 network round-trip time** for the access to the preferred (close) **chunk server**.

While the round-trip time to the data storage itself can be reduced by having duplicates at a close-by chunkserver, that is not possible with the metadata.  I have to choose the one or other location for running `mfsmaster`.

I was kind of hoping that some (client) cache option would allow the client to not have to do the metadata lookup for files that are in the same directory as recently accessed files.  Or that recently accessed files themselves would have much smaller access times.  These are the results from a second repeated access (directly after a first access):

| dir  | read from node A | read from node B |
| ---- | ----------------------- | ----------------------- |
| A   |  0.2 ms/file           | **43 ms/file**              |
| B   |  **0.3  ms/file**            | 43  ms/file             |
| AB | 0.2  ms/file           |  33 ms/file            |

The data access penalty is now gone (especially visible in the **bold** items).

Is there some option I have missed that could also reduce the metadata access times by using cached entries and avoid the network round-trip?

(The tests were done with the Debian stable/bullseye packages of moosefs (3.0.115-1).)

**Comment by chogata** [comment]
First, MooseFS client does not cache metadata* - it's the kernel's job. The timeouts are what MooseFS tells the kernel, when the kernel is asking about a file: "here is the file's metadata and you are allowed to remember them for no longer than this time". So yes, every first access to any file requires the roundtrip to the master. After that it depends on the kernel. 
Second, MooseFS does not pre-fetch metadata (as opposed to data, which it does in files read sequentially). There is no way to predict which metadata might be useful in the future and pre-fetching metadata that might be absolutely un-necessary is pointless and resource consuming. In your scenario it would help, but in others - like reading 1 file from a directory containing 100 000 of files - it would be a big disadvantage. And MooseFS is used in all kinds of environments.

BTW: this option is quite dangerous: `-o mfsnegentrycacheto=100` and usually not needed, so I would suggest not changing the default (0) even when you change all the other "cache" ones.

*ok, it does cache a little, in that a `lookup` will fetch any information that might be needed for a subsequent `open` for a regular file, therefore there is one round-trip instead of 2 or 3 that might be needed for a lookup&open combo; imagine your tests without that :)

**Reply by inkdot7** [reply]
Thanks for the explanation!

(I understand that some of the options I tried are not healthy, but idea was to first find out what may can change the performance characteristics, and then consider if they are reasonable to use.)

How would one change the kernel metadata cache behaviour?  (My searches so far in finding any example has not been successful.)

It looks like the fuse options [kernel_cache](https://manpages.debian.org/bullseye/fuse/mount.fuse.8.en.html#kernel_cache) (I know this is not for safe network file systems), [entry_timeout=T](https://manpages.debian.org/bullseye/fuse/mount.fuse.8.en.html#entry_timeout=T) and [attr_timeout=T](https://manpages.debian.org/bullseye/fuse/mount.fuse.8.en.html#attr_timeout=T) could be interesting to try, but doing e.g. `mfsmount /mfs -o attr_timeout=19` just gives:
```
mfsmaster accepted connection with parameters: read-write,restricted_ip,admin ; root mapped to root:root
fuse: unknown option(s): `-o attr_timeout=19'
error in fuse_session_new
```

Also adding the options to `/etc/fuse.conf` had no effect.
